### PR TITLE
[AMDGPU] promote i1 arg type for amdgpu_cs

### LINKED ISSUE #68087: i1 arguments do not work with amdgpu_cs
```
; RUN: llc -march=amdgcn -mcpu=gfx900 < %s
define amdgpu_cs void @main(i1 %i83) {
bb:
  ret void
}

```

LLVM ERROR: unable to allocate function argument #0

### PR DESCRIPTION
**Pull request by nickleus27**
fixes #68087 
Not sure where to put regression tests for this pr? Also, should i1 args not in reg also be promoted?

### CODE REVIEW

**Comment by arsenm** [comment]
> fixes #68087 Not sure where to put regression tests for this pr? Also, should i1 args not in reg also be promoted?

test/CodeGen/AMDGPU/calling-conventions.ll is probably the best bet right now (although I'm not happy with how these tests are organized)

Both inreg and not inreg should be promoted. The inreg just flips from VGPR to SGPR.

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 1]
```diff
@@ -2078,4 +2078,13 @@ entry:
...
 
+; test that i1 argument for amdgpu_cs is working
+define amdgpu_cs void @amdgpu_cs_i1_arg(i1 %arg0) {
+; GCN-LABEL: amdgpu_cs_i1_arg:
+; GCN:       ; %bb.0: ; %bb
+; GCN-NEXT:    s_endpgm
+bb:
+  ret void
+}
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2089]: Can you also check the inreg case? Also some i1 vectors

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 2]
```diff
@@ -2078,4 +2078,13 @@ entry:
   ret void
 }
 
+; test that i1 argument for amdgpu_cs is working
+define amdgpu_cs void @amdgpu_cs_i1_arg(i1 %arg0) {
+; GCN-LABEL: amdgpu_cs_i1_arg:
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2083]: An unused argument doesn't show much useful, should add some sample use so we can see the register used

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 3]
```diff
@@ -2078,12 +2078,219 @@ entry:
...
+; SI:       ; %bb.0:
+; SI-NEXT:    s_mov_b32 s3, 0xf000
+; SI-NEXT:    s_mov_b32 s2, -1
+; SI-NEXT:    v_mov_b32_e32 v0, 0
+; SI-NEXT:    buffer_store_byte v0, off, s[0:3], 0
+; SI-NEXT:    s_endpgm
+;
+; VI-LABEL: amdgpu_cs_inreg_i1:
+; VI:       ; %bb.0:
+; VI-NEXT:    v_mov_b32_e32 v0, 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2200]: This is brokenly just storing a constant 0

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 4]
```diff
@@ -2078,12 +2078,219 @@ entry:
...
+; VI-NEXT:    s_endpgm
+;
+; GFX11-LABEL: amdgpu_cs_inreg_i1:
+; GFX11:       ; %bb.0:
+; GFX11-NEXT:    v_mov_b32_e32 v0, 0
+; GFX11-NEXT:    global_store_b8 v[0:1], v0, off
+; GFX11-NEXT:    s_nop 0
+; GFX11-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
+; GFX11-NEXT:    s_endpgm
+  %add = add i1 %arg0, %arg0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2211]: add i1 probably not the best test, just store the value directly?

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 5]
```diff
@@ -2078,12 +2078,219 @@ entry:
...
+; SI:       ; %bb.0:
+; SI-NEXT:    s_mov_b32 s3, 0xf000
+; SI-NEXT:    s_mov_b32 s2, -1
+; SI-NEXT:    v_mov_b32_e32 v0, 0
+; SI-NEXT:    buffer_store_byte v0, off, s[0:3], 0
+; SI-NEXT:    s_endpgm
+;
+; VI-LABEL: amdgpu_cs_i1:
+; VI:       ; %bb.0:
+; VI-NEXT:    v_mov_b32_e32 v0, 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2092]: Brokenly storing a constant 0

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 6]
```diff
@@ -2078,12 +2078,219 @@ entry:
...
+; VI-NEXT:    s_endpgm
+;
+; GFX11-LABEL: amdgpu_cs_i1:
+; GFX11:       ; %bb.0:
+; GFX11-NEXT:    v_mov_b32_e32 v0, 0
+; GFX11-NEXT:    global_store_b8 v[0:1], v0, off
+; GFX11-NEXT:    s_nop 0
+; GFX11-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
+; GFX11-NEXT:    s_endpgm
+  %add = add i1 %arg0, %arg0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2103]: these adds all fold to 0, so these still aren't showing the argument usage

**Comment by rovka** [comment]
Hi, while you're at it, could you please make the same change for [`amdgpu_cs_chain` and `amdgpu_cs_chain_preserve`](https://github.com/llvm/llvm-project/blob/6c7aee0ba51a45dc387d7da6723857bba0c5587c/llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td#L227)?

File: llvm/test/CodeGen/AMDGPU/calling-conventions.ll [hunk 7]
```diff
@@ -2078,4 +2078,1169 @@ entry:
   ret void
 }
 
+define amdgpu_cs void @amdgpu_cs_i1(i1 %arg0) {
+; SI-LABEL: amdgpu_cs_i1:
+; SI:       ; %bb.0:
+; SI-NEXT:    v_and_b32_e32 v0, 1, v0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/calling-conventions.ll:2084]: Could also test the signext/zeroext cases, in which cases we can avoid the instruction

**Review by arsenm — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Pull request merged**